docs/decisions/backend/0011-advanced-installer-persistence.md

```diff
@@ -0,0 +1,147 @@
+# Persistence/Restore Functionality for Advanced Installer
+
+Manual installs need to be remembered across different deployments to avoid the user having to remap them every time
+between installations.
+
+## Context and Problem Statement
+
+After manually installing a mod via the Advanced Installer, the application needs to save the details of the user's
+last set of choices made in the Installer.
+
+This is such that, when the application needs to perform a re-deployment occurs, the end user does not have to go through
+the manual installation steps again.
+
+Note: Re-deployment can happen due to a variety of factors, such as a mod deployed before the Advanced Installed mod
+being removed from a loadout. Not all re-deployments are due to the user's direct actions/expectations. Therefore a
+persistence mechanism here is crucial.
```

Comment:
I don't understand, why is this needed?
Don't we already store the archive source and destination in IToFile as a GamePath, what do we need more?

Reply:
It's because of the inherent design of our Data Store.

Suppose we have a loadout of mods like the following (arrows indicate order in which they were added):

```mermaid
flowchart LR
    A[Mod A] --> B[Mod B]
    B --> C[FomodOrAdvanced]
    C --> D[Mod D]
```

Where `FomodOrAdvanced` is our mod in question.

If the user removes `Mod B`, the DataStore performs a rollback to before `Mod B` was added, so the flowchart becomes

```mermaid
flowchart LR
    A[Mod A]
```

and then, the remaining items are then freshly 're-deployed' from scratch on top of `Mod A`.

```mermaid
flowchart LR
    A[Mod A]  --> C[FomodOrAdvanced]
    C --> D[Mod D]
```

This means that for all mods after `Mod A`, we have to drop existing information and re-run all of the installers from scratch (i.e. this accommodates for things like FOMOD file based dependencies). 

So for `FomodOrAdvanced` and `Mod D`, we end up having to run all of the installers again (`GetModsAsync`); which means that the user will have to make their Advanced Installer (and/or FOMOD) choices from scratch again.

Reply:
No, we don't ask users to re-run installers in this case. 
I think we need to have a discussion about this because I don't think we are on the same page on how the main app should handle mods and deployment.

Each installed mod is a collection of file sources and destinations. The source will be a .NX archive containing all the downloaded mod files, destination is where the file needs to be deployed.

If a user removes a mod this doesn't really influence all the other mods installed after it. Each mod is an independent entity. 
Deployment phase just takes the list of files from each enabled mod, if there are two mods wanting to install different files to the same file path, then we solve this using some user defined priority or conflict resolution rules.

This is all virtual, then when the user wants to actually play the game, we try to match the filesystem state with our virtual loadout state (Adding new files, removing files that don't appear in the current loadout (backing them up in NX archive if we don't already have their hash in the db).

Reply:
Seems you're right, my knowledge of the internals is outdated in this scenario. I had a walk through the source to confirm.
I'll adjust the PR accordingly.

Reply:
Resolved in ADR.